File: android/app/src/main/java/ch/subri/morninglight/data/entity/Alarm.kt
package ch.subri.morninglight.data.entity

import androidx.room.ColumnInfo
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity
data class Alarm(
    @PrimaryKey(autoGenerate = true) val id: Long,
    val title: String,
    @Embedded val time: Time,
    val duration: Int,
    @ColumnInfo(defaultValue = "FALSE") val enabled: Boolean,
    @Embedded val activity: AlarmActivity,
)

fun Alarm.overlapsWith(other: Alarm): Boolean = activity.overlapsWith(other.activity)

fun Alarm.overlapsWith(others: List<Alarm>): Boolean = others
    .map { this.overlapsWith(it) }
    .any { it }

File: arduino/lib/AlarmManager/src/Alarm.cpp
#include <Alarm.h>

Alarm::Alarm() : Alarm(false, Time(), 1) {}

Alarm::Alarm(bool enabled, Time startTime, unsigned int duration)
    : enabled(enabled), startTime(startTime), duration(duration) {}

bool Alarm::equals(const Alarm& other) const {
    return this->duration == other.duration && this->enabled == other.enabled &&
           this->startTime == other.startTime;
}